adr/0025-composition.md

```diff
@@ -0,0 +1,282 @@
+    charts:
+      - name: uds-mattermost-config
+        namespace: mattermost
+        version: 0.1.0
+        localPath: ./chart
+        valuesFiles:
+          - values/config-values.yaml
+      - name: mattermost-enterprise-edition
+        namespace: mattermost
+        url: https://helm.mattermost.com
+        gitPath: chart
+        version: 2.6.55
+        valuesFiles:
+          - values/common-values.yaml
+
+          ## No longer needed because this was only used for replacing
+          ## image tags, see new `images` API usage below.
+          #
+          # - values/upstream-values.yaml
+
+    ## Kustomize-style image replacements facilitate proper composition.
+    ## This makes it easy for downstream components to override image tags
+    ## without knowing anything about the Helm chart(s) being referenced
+    ## nor their chart-specific supported `values`.
+    #
+    # images:
+    #   - appropriate/curl:latest
+    #   - mattermost/mattermost-enterprise-edition:9.9.0
+    images:
+      - name: appropriate/curl
+        newTag: latest
+      - name: mattermost/mattermost-enterprise-edition
+        newTag: 9.9.0
+    actions:
+      onDeploy:
+        after:
+          - description: Validate Mattermost Package
+            maxTotalSeconds: 300
+            wait:
+              cluster:
+                kind: Packages
+                name: mattermost
+                namespace: mattermost
+                condition: "'{.status.phase}'=Ready"
+
+  - name: mattermost-registry1
+
+    ## Previously a `required` component with the "registry1" `flavor`,
+    ## `mattermost-registry1` is now defined simply as a component that directly
+    ## extends the default `mattermost` component above.
+    #
+    # required: true
+    # import:
+    #   path: common
+    # only:
+    #   flavor: registry1
+    #   cluster:
+    #     architecture: amd64
+    extends: mattermost
+    required: false
+
+    ## In this case (and as is the case with most `registry1` component flavors)
+    ## the Helm chart values were only used to override image tags. We can now
+    ## take advantage of a more robust and declarative `images` replacements API.
+    #
+    # charts:
+    #   - name: mattermost-enterprise-edition
+    #     valuesFiles:
+    #       - values/registry1-values.yaml
+    #       ## sample contents of `values/registry1-values.yaml`
+    #       # mattermostApp:
+    #       #   image:
+    #       #     repository: registry1.dso.mil/ironbank/opensource/mattermost/mattermost
+    #       #     tag: 9.9.0
+    #       # initContainerImage:
+    #       #   repository: registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal
+    #       #   tag: 9.4
+    # images:
+    #   - registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal:9.4
+    #   - registry1.dso.mil/ironbank/opensource/mattermost/mattermost:9.9.0
+    images:
+      - name: appropriate/curl
+        newName: registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal
```

Comment:
feels verbose / complex to me (you'd have to match strings in a Helm post render which might not be reliable - these references could be anywhere including a command arg passed into an operator)

Reply:
Verbose relative to what? The proposed API allows you to avoid specifying Helm chart value overrides entirely while still declaring which images to include in the package.

> you'd have to match strings in a Helm post render which might not be reliable

This should be performed as structural edits on the YAML, not by string matching. Here are the field specs that `kustomize` looks at by default: https://github.com/kubernetes-sigs/kustomize/blob/c1de0301f5e71ffbecf77f5bc8687e7693878eb5/api/internal/konfig/builtinpluginconsts/images.go#L9-L16